repo: add install instructions, Makefile to install dependencies and run unit tests

cmoussa1 opened the pull request:
Noted by @dongahn in #6 and #12, the install and unit test instructions in the **README.md** were not clear/non-existent. This PR adds a **Makefile** that installs the dependencies listed in **requirements.txt** as well as runs the unit tests, so both can be run with `make check`.

```
$ make check
pip3 install -r requirements.txt
Collecting pandas==0.25.3
Installing collected packages: pandas
Successfully installed pandas-0.25.3
python3 -m unittest discover
....
----------------------------------------------------------------------
Ran 4 tests in 0.021s

OK
```

**README.md** has also been updated to reflect these instructions. 

Fixes #12

dongahn commented on README.md:
```diff
@@ -2,14 +2,26 @@ _NOTE: The interfaces of flux-accounting are being actively developed and are no
+##### Install instructions
 
-Running as a cron job, `write_jobs.py` uses a flux-core RPC to retrieve job data that have recently transitioned to an INACTIVE state in the past _x_ minutes. It parses the job record data and writes it to a SQLite database, which can be queried at a later time. 
+This repository uses `pandas` for data display information. To install this dependency, along with any other packages specified in **requirements.txt**, and also run unit tests, you can run `make check`:
```
Thanks @cmoussa1. Strictly speaking, run tests and make check are not part of Install instructions? 

Would it make sense to have "build instructions", "test instructions" and "install instructions"? And discuss the step for each in different section? 

Is `pandas` a test program dependency or a dependency for the core flux-accounting?

cmoussa1 replied:
Thanks for the feedback @dongahn. I believe what's required as far as install instructions after pulling down this repository are the following (let me know if you had to do anything else on your end):

- have `python3` installed in your environment (I used `brew install python3`)
- have `pip3` installed in your environment (which should be included when installing `python3`)
- install the dependencies listed in **requirements.txt** (which for now is just `pandas`) with `pip3 install -r requirements.txt`

`pandas` is a dependency for the core flux-accounting.

dongahn commented on README.md:
````diff
@@ -2,14 +2,26 @@ _NOTE: The interfaces of flux-accounting are being actively developed and are no
+This repository uses `pandas` for data display information. To install this dependency, along with any other packages specified in **requirements.txt**, and also run unit tests, you can run `make check`:
 
-##### user account information
+```
````
Are there any other dependencies? Does yours depend on flux-core or flux-sched at this point? I think I saw one of the module depends on flux-core's python binding. If so, do you want to say "flux-sched requires an installed flux-core package"?

cmoussa1 replied:
Thanks for catching that. At this point, flux-accounting doesn't depend on flux-core's python binding. I was going to decorate the main function with flux-core's `util.CLIMain` decorator, but I decided to just include my own logger for now, which uses Python's standard `logging` module.

dongahn replied:
Ah. Cool. This can make the testing of your core logic easier!

dongahn commented on README.md:
````diff
@@ -2,14 +2,26 @@ _NOTE: The interfaces of flux-accounting are being actively developed and are no
 
-##### user account information
+```
+$ make check
````
I am not sure if I don't know if adding Makefile support for `make check` in an ad hoc fashion would be a long term solution. Did you add this as a short term stopgap?

@SteVwonder or @trws: is there any good python project that we can model after in terms of how to build, how to test and how to install etc?

SteVwonder replied:
I think this python guide is a good data point, in particular the section on using a Makefile for kicking off the dependency installation and the test suite: https://docs.python-guide.org/writing/structure/#makefile

The python setuptools documentation mentions that the `test` command is deprecated and recommends the python package tox: https://setuptools.readthedocs.io/en/latest/setuptools.html#test-build-package-and-run-a-unittest-suite.  I think that tox is probably overkill right now for this project.

In terms of installation and distribution, `setup.py sdist` will create a source distribution that can be uploaded to PyPI.  `setup.py install` and `pip install .` will both install locally from the source tree.  `setup.py develop` and `pip install -e .` will install into develop mode (makes it look installed, but you can see edits to source files without having to re-install).  Relevant Docs:
- https://packaging.python.org/tutorials/installing-packages/
- https://setuptools.readthedocs.io/en/latest/setuptools.html#development-mode

dongahn replied:
Thanks. 

At first glance, it looks like autotools aren't best practices for python projects...

It will be a good test whether we can install our flux-accounting package properly with `flux-core` and `flux-sched` packages including support for both co-installation and standalone  installation) without GNU autotools and testing these configurations

`flux-sched` benefited heavily from mirroring various GNU autotools features.

dongahn commented on README.md:
````diff
@@ -2,14 +2,26 @@ _NOTE: The interfaces of flux-accounting are being actively developed and are no
-##### user account information
+```
+$ make check
+pip3 install -r requirements.txt
````
When I used flux-sched docker, pip3 wasn't there so I had to manually install it. So you may as well mention pip3 as a dependency.

cmoussa1 commented:
Thanks for your suggestions and feedback @SteVwonder and @dongahn - I've gone ahead and pushed a couple of changes:

- I've added two Dockerfiles to this repo, one for `centos8` and one for `ubuntu`. As of now, they sit in the top level directory and are launched with the following command:

`docker build -f Dockerfile.ubuntu .`

I've only been able to get them to work when they are both sitting in the top level directory. Is it better to have them sitting in a `docker_files/` directory or something? If it is, I'm unsure how to structure it. 

- I added the dependency table to **README.md** for ubuntu and centos8, which includes `python3.6`, `pip3`, and `tox`.

> At first glance, it looks like autotools aren't best practices for python projects...

Okay! For now I have left the `Makefile` in this branch and added the Python-specific build and install commands to **README.md**. Would you prefer we leave the Makefile out of this repo entirely?

dongahn commented:
> I've only been able to get them to work when they are both sitting in the top level directory. Is it better to have them sitting in a docker_files/ directory or something? If it is, I'm unsure how to structure it.

How about mirroring `flux-core` and `flux-sched` and put those Dockerfiles in new directories 

- `src/test/docker/bionic`
- `src/test/docker/centos7`
- `src/test/docker/centors8`


This way, you will be able to reuse many of travis testing infrastructure from these repos. 

You won't have much inside `src` directories but my guess is it will grow, going forward. As you add front end commands like reporting tool, it will likely go into `src/cmd`.

cmoussa1 commented:
Thanks @dongahn. I went ahead and moved the Dockerfiles to their respective directories inside `accounting`, so the path is: `flux-accounting/accounting/test/docker/<os>/Dockerfile.<os`, and they should each be able to be run with: `docker build -f <file name> ../../../../`.

dongahn commented on README.md:
```diff
@@ -2,14 +2,43 @@ _NOTE: The interfaces of flux-accounting are being actively developed and are no
+### Test Instructions
 
-Users have banks that they can charge their jobs against; this will affect their priority when submitting future jobs in order to maintain a fair balance of resources between users running jobs on a single cluster. Multiple factors play a role in determining a user's job priority, but in general, the amount of resources used vs. the resource limits initially allocated will either lower or heighte ...
+Make sure all is right correctly by running the unit tests with `tox`:
```
< Make sure all is right correctly by running the unit tests with `tox`:

> Run the unit tests with `tox` to ensure the correctness of this package on your platform:

cmoussa1 replied:
Thanks @dongahn 👍 I went ahead and pushed a change to reflect your suggestion.

dongahn commented:
@cmoussa1: I just made one minor editing suggestion. Once that's addressed, this can go in from my perspective.

cmoussa1 commented:
Thanks!
